File: class27/java/TCC2701-MiniCompilador.java
// TCC2701-MiniCompilador.java | Aula 27
// Esqueleto do Projeto Integrador A3 — mini-interpretador.
// Linguagem: variáveis, expressões, if/else, while, print.
// Espelho Java do TCC2701-MiniCompilador.py (mesma linguagem e saídas).
import java.util.*;

public class TCC2701MiniCompilador {

    // ---------- Token ----------
    static class Token {
        String tipo, valor; int linha, col;
        Token(String t, String v, int l, int c) { tipo = t; valor = v; linha = l; col = c; }
    }

    // ---------- Scanner ----------
    static List<Token> scanner(String fonte) {
        List<Token> tokens = new ArrayList<>();
        int i = 0, n = fonte.length(), linha = 1, col = 1;
        Set<String> keywords = new HashSet<>(Arrays.asList("if", "else", "while", "print"));
        while (i < n) {
            char c = fonte.charAt(i);
            if (c == '\n') { i++; linha++; col = 1; continue; }
            if (Character.isWhitespace(c)) { i++; col++; continue; }
            if (c == '#') { while (i < n && fonte.charAt(i) != '\n') i++; continue; }
            int startCol = col;
            if (Character.isDigit(c)) {
                int j = i; while (j < n && Character.isDigit(fonte.charAt(j))) j++;
                if (j < n && fonte.charAt(j) == '.') { j++; while (j < n && Character.isDigit(fonte.charAt(j))) j++; }
                tokens.add(new Token("NUM", fonte.substring(i, j), linha, startCol));
                col += j - i; i = j; continue;
            }
            if (Character.isLetter(c) || c == '_') {
                int j = i; while (j < n && (Character.isLetterOrDigit(fonte.charAt(j)) || fonte.charAt(j) == '_')) j++;
                String lex = fonte.substring(i, j);
                tokens.add(new Token(keywords.contains(lex) ? "KEYWORD" : "ID", lex, linha, startCol));
                col += j - i; i = j; continue;
            }
            if ("+-*/=<>".indexOf(c) >= 0) {
                if (c == '=' && i + 1 < n && fonte.charAt(i + 1) == '=') {
                    tokens.add(new Token("OP", "==", linha, startCol)); i += 2; col += 2; continue;
                }
                tokens.add(new Token("OP", String.valueOf(c), linha, startCol)); i++; col++; continue;
            }
            if ("(){};".indexOf(c) >= 0) {
                tokens.add(new Token("PUNCT", String.valueOf(c), linha, startCol)); i++; col++; continue;
            }
            throw new RuntimeException("Caractere inválido: '" + c + "' linha " + linha + " col " + col);
        }
        tokens.add(new Token("EOF", "", linha, col));
        return tokens;
    }

    // ---------- AST ----------
    static abstract class No {}
    static class Num extends No { double val; Num(double v) { val = v; } }
    static class Var extends No { String nome; Var(String n) { nome = n; } }
    static class BinOp extends No { String op; No esq, dir; BinOp(String o, No e, No d) { op = o; esq = e; dir = d; } }
    static class Assign extends No { String nome; No expr; Assign(String n, No e) { nome = n; expr = e; } }
    static class Print extends No { No expr; Print(No e) { expr = e; } }
    static class If extends No { No cond; List<No> thenB, elseB; If(No c, List<No> t, List<No> e) { cond = c; thenB = t; elseB = e; } }
    static class While extends No { No cond; List<No> body; While(No c, List<No> b) { cond = c; body = b; } }

    // ---------- Parser ----------
    static class Parser {
        List<Token> tokens; int pos = 0;
        Parser(List<Token> t) { tokens = t; }
        Token peek() { return tokens.get(pos); }
        Token consume() { return tokens.get(pos++); }
        Token consume(String tipo) {
            Token t = tokens.get(pos);
            if (!t.tipo.equals(tipo)) throw new RuntimeException("Esperado " + tipo + ", obtido " + t.tipo + "('" + t.valor + "') linha " + t.linha);
            pos++; return t;
        }
        Token consume(String tipo, String valor) {
            Token t = consume(tipo);
            if (!t.valor.equals(valor)) throw new RuntimeException("Esperado '" + valor + "', obtido '" + t.valor + "' linha " + t.linha);
            return t;
        }

        List<No> parse() {
            List<No> stmts = new ArrayList<>();
            while (!peek().tipo.equals("EOF")) stmts.add(stmt());
            return stmts;
        }

        No stmt() {
            Token t = peek();
            if (t.tipo.equals("KEYWORD") && t.valor.equals("if")) return ifStmt();
            if (t.tipo.equals("KEYWORD") && t.valor.equals("while")) return whileStmt();
            if (t.tipo.equals("KEYWORD") && t.valor.equals("print")) return printStmt();
            if (t.tipo.equals("ID")) return assign();
            throw new RuntimeException("Token inesperado: " + t.tipo + "('" + t.valor + "')");
        }

        No assign() {
            String nome = consume("ID").valor;
            consume("OP", "=");
            No expr = expr();
            consume("PUNCT", ";");
            return new Assign(nome, expr);
        }

        No ifStmt() {
            consume("KEYWORD", "if"); consume("PUNCT", "(");
            No cond = expr(); consume("PUNCT", ")");
            List<No> thenB = block();
            List<No> elseB = new ArrayList<>();
            if (peek().tipo.equals("KEYWORD") && peek().valor.equals("else")) {
                consume("KEYWORD", "else"); elseB = block();
            }
            return new If(cond, thenB, elseB);
        }

        No whileStmt() {
            consume("KEYWORD", "while"); consume("PUNCT", "(");
            No cond = expr(); consume("PUNCT", ")");
            return new While(cond, block());
        }

        No printStmt() {
            consume("KEYWORD", "print"); consume("PUNCT", "(");
            No expr = expr(); consume("PUNCT", ")"); consume("PUNCT", ";");
            return new Print(expr);
        }

        List<No> block() {
            consume("PUNCT", "{");
            List<No> stmts = new ArrayList<>();
            while (!peek().valor.equals("}")) stmts.add(stmt());
            consume("PUNCT", "}");
            return stmts;
        }

        No expr() { return binop(term(), "+-"); }
        No term() { return binop(factor(), "*/"); }

        No binop(No esq, String ops) {
            while (peek().tipo.equals("OP") && ops.contains(peek().valor)) {
                String op = consume().valor;
                No dir = ops.contains("+") ? term() : factor();
                esq = new BinOp(op, esq, dir);
            }
            return esq;
        }

        No factor() {
            Token t = peek();
            if (t.tipo.equals("NUM")) { consume(); return new Num(Double.parseDouble(t.valor)); }
            if (t.tipo.equals("ID")) { consume(); return new Var(t.valor); }
            if (t.valor.equals("(")) { consume("PUNCT", "("); No e = expr(); consume("PUNCT", ")"); return e; }
            throw new RuntimeException("Token inesperado em factor: " + t.tipo + "('" + t.valor + "')");
        }
    }

    // ---------- Interpretador ----------
    static class Interpretador {
        Map<String, Double> vars = new HashMap<>();

        void executar(List<No> stmts) { for (No s : stmts) execStmt(s); }

        void execStmt(No s) {
            if (s instanceof Assign) { Assign a = (Assign) s; vars.put(a.nome, eval(a.expr)); }
            else if (s instanceof Print) { System.out.println(eval(((Print) s).expr)); }
            else if (s instanceof If) { If i = (If) s; if (eval(i.cond) != 0) executar(i.thenB); else executar(i.elseB); }
            else if (s instanceof While) { While w = (While) s; while (eval(w.cond) != 0) executar(w.body); }
        }

        double eval(No e) {
            if (e instanceof Num) return ((Num) e).val;
            if (e instanceof Var) {
                Var v = (Var) e;
                if (!vars.containsKey(v.nome)) throw new RuntimeException("Variável '" + v.nome + "' não definida");
                return vars.get(v.nome);
            }
            if (e instanceof BinOp) {
                BinOp b = (BinOp) e;
                double esq = eval(b.esq), dir = eval(b.dir);
                switch (b.op) {
                    case "+": return esq + dir;
                    case "-": return esq - dir;
                    case "*": return esq * dir;
                    case "/": return esq / dir;
                    case ">": return esq > dir ? 1 : 0;
                    case "<": return esq < dir ? 1 : 0;
                    default: throw new RuntimeException("Operador inválido: " + b.op);
                }
            }
            throw new RuntimeException("Nó não suportado: " + e.getClass().getSimpleName());
        }
    }

    public static void main(String[] args) {
        String fonte = "x = 3 + 4 * 2;\nprint(x);\nif (x > 10) {\n    print(1);\n} else {\n    print(0);\n}\n";
        System.out.println("=== Fonte ===");
        System.out.println(fonte);
        List<Token> tokens = scanner(fonte);
        List<No> ast = new Parser(tokens).parse();
        System.out.println("=== Execução ===");
        new Interpretador().executar(ast);
    }
}

File: class27/python/TCC2701-MiniCompilador.py
#!/usr/bin/env python3
"""TCC2701-MiniCompilador.py | Aula 27
Esqueleto do Projeto Integrador A3 — mini-interpretador.
Linguagem: variáveis, expressões, if/else, while, print.
Stack: Python puro (sem dependências) — pode ser portado para ANTLR/PLY.
"""
import re
from typing import Dict, List, Tuple


# ---------- Scanner ----------
TOKEN_SPEC = [
    ("NUM",     r"\d+(\.\d+)?"),
    ("ID",      r"[a-zA-Z_]\w*"),
    ("OP",      r"==|[-+*/=<>]"),
    ("PUNCT",   r"[(){};]"),
    ("WS",      r"\s+"),
    ("COMMENT", r"#[^\n]*"),
]
KEYWORDS = {"if", "else", "while", "print"}


def scanner(fonte: str) -> List[Tuple[str, str, int, int]]:
    tokens = []
    pos = 0
    linha = 1
    col = 1
    while pos < len(fonte):
        for tipo, pattern in TOKEN_SPEC:
            m = re.match(pattern, fonte[pos:])
            if m:
                valor = m.group(0)
                if tipo == "WS":
                    col += len(valor)
                    pos += len(valor)
                    break
                if tipo == "COMMENT":
                    pos += len(valor)
                    break
                if tipo == "ID" and valor in KEYWORDS:
                    tipo = "KEYWORD"
                tokens.append((tipo, valor, linha, col))
                col += len(valor)
                pos += len(valor)
                break
        else:
            if fonte[pos] == "\n":
                linha += 1
                col = 1
                pos += 1
            else:
                raise SyntaxError(f"Caractere inválido {fonte[pos]!r} linha {linha} col {col}")
    tokens.append(("EOF", "", linha, col))
    return tokens


# ---------- Parser (descendente recursivo) ----------
class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0

    def peek(self):
        return self.tokens[self.pos]

    def consume(self, tipo=None, valor=None):
        t = self.tokens[self.pos]
        if tipo and t[0] != tipo:
            raise SyntaxError(f"Esperado {tipo}, obtido {t[0]}({t[1]!r}) linha {t[2]}")
        if valor and t[1] != valor:
            raise SyntaxError(f"Esperado {valor!r}, obtido {t[1]!r} linha {t[2]}")
        self.pos += 1
        return t

    def parse(self):
        stmts = []
        while self.peek()[0] != "EOF":
            stmts.append(self.stmt())
        return stmts

    def stmt(self):
        t = self.peek()
        if t[0] == "KEYWORD" and t[1] == "if":
            return self.if_stmt()
        if t[0] == "KEYWORD" and t[1] == "while":
            return self.while_stmt()
        if t[0] == "KEYWORD" and t[1] == "print":
            return self.print_stmt()
        if t[0] == "ID":
            return self.assign()
        raise SyntaxError(f"Token inesperado: {t}")

    def assign(self):
        nome = self.consume("ID")[1]
        self.consume("OP", "=")
        expr = self.expr()
        self.consume("PUNCT", ";")
        return ("assign", nome, expr)

    def if_stmt(self):
        self.consume("KEYWORD", "if")
        self.consume("PUNCT", "(")
        cond = self.expr()
        self.consume("PUNCT", ")")
        then_block = self.block()
        else_block = []
        if self.peek()[0] == "KEYWORD" and self.peek()[1] == "else":
            self.consume("KEYWORD", "else")
            else_block = self.block()
        return ("if", cond, then_block, else_block)

    def while_stmt(self):
        self.consume("KEYWORD", "while")
        self.consume("PUNCT", "(")
        cond = self.expr()
        self.consume("PUNCT", ")")
        body = self.block()
        return ("while", cond, body)

    def print_stmt(self):
        self.consume("KEYWORD", "print")
        self.consume("PUNCT", "(")
        expr = self.expr()
        self.consume("PUNCT", ")")
        self.consume("PUNCT", ";")
        return ("print", expr)

    def block(self):
        self.consume("PUNCT", "{")
        stmts = []
        while self.peek()[1] != "}":
            stmts.append(self.stmt())
        self.consume("PUNCT", "}")
        return stmts

    def expr(self):
        return self.expr_binop(self.term, ["+", "-"])

    def term(self):
        return self.expr_binop(self.factor, ["*", "/"])

    def factor(self):
        t = self.peek()
        if t[0] == "NUM":
            self.consume()
            return ("num", float(t[1]))
        if t[0] == "ID":
            self.consume()
            return ("var", t[1])
        if t[1] == "(":
            self.consume("PUNCT", "(")
            e = self.expr()
            self.consume("PUNCT", ")")
            return e
        raise SyntaxError(f"Token inesperado em factor: {t}")

    def expr_binop(self, sub, ops):
        esq = sub()
        while self.peek()[0] == "OP" and self.peek()[1] in ops:
            op = self.consume()[1]
            dir_ = sub()
            esq = ("binop", op, esq, dir_)
        return esq


# ---------- Interpretador ----------
class Interpretador:
    def __init__(self):
        self.vars: Dict[str, float] = {}

    def executar(self, stmts):
        for s in stmts:
            self.exec_stmt(s)

    def exec_stmt(self, s):
        if s[0] == "assign":
            val = self.eval(s[2])
            self.vars[s[1]] = val
        elif s[0] == "print":
            print(self.eval(s[1]))
        elif s[0] == "if":
            if self.eval(s[1]) != 0:
                self.executar(s[2])
            else:
                self.executar(s[3])
        elif s[0] == "while":
            while self.eval(s[1]) != 0:
                self.executar(s[2])

    def eval(self, e):
        if e[0] == "num":
            return e[1]
        if e[0] == "var":
            if e[1] not in self.vars:
                raise NameError(f"Variável '{e[1]}' não definida")
            return self.vars[e[1]]
        if e[0] == "binop":
            esq = self.eval(e[2])
            dir_ = self.eval(e[3])
            ops = {"+": esq + dir_, "-": esq - dir_, "*": esq * dir_, "/": esq / dir_}
            return ops[e[1]]


if __name__ == "__main__":
    fonte = """
x = 3 + 4 * 2;
print(x);
if (x > 10) {
    print(1);
} else {
    print(0);
}
"""
    print("=== Fonte ===")
    print(fonte)
    tokens = scanner(fonte)
    ast = Parser(tokens).parse()
    print("=== Execução ===")
    Interpretador().executar(ast)
